### DISCUSSION TITLE
Unusual Memory Usage Issue with MOOSE InversePowerMethod Executioner 'LINEAR' Solver

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#26419] [Q&A General]
I have developed an SN-DFEM program using MOOSE, and recently, I've been experimenting with the 'LINEAR' solver of the InversePowerMethod executioner. However, I've encountered a peculiar phenomenon: initially, I tested it on a one-dimensional problem with 800,000 degrees of freedom, and according to the system monitor, MOOSE occupied around 700MB of memory. However, when I tested it on a three-dimensional problem with only 80,000 degrees of freedom, the system monitor showed that MOOSE reached a maximum memory usage of around 9.9GB. Conversely, when I used PJFNK to solve the three-dimensional problem, MOOSE only occupied approximately 200MB of memory.

My understanding of the LINEAR solver is that it solves a linear problem Ax=b where b=Bx^(n-1)/λ^(n-1) using the krylov method (not JFNK) and requires the complete solution of the entire Jacobian matrix. Storing the Jacobian matrix can indeed be memory-intensive, but the memory usage observed in the tests seems excessively large and unreasonable.

Could anyone shed light on what might be causing this issue? Additionally, where can I find the code for the LINEAR solver of the InversePowerMethod executioner?

Thank you in advance for any insights.

**Comment by GiudGiud** [comment]
Hello

If you use Newton or PJFNK with 
```
[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]
```
do you see the same memory utilization as LINEAR?

In 3D, are you using a hex mesh or a tetrahedral mesh?

Guillaume

**Reply by hityyds** [reply]
Oh, yes! When I use the SMP preconditioner with either the Newton or PJFNK methods, MOOSE exhibits the same memory usage behavior. In 3D, I used a cuboidal mesh. While I've identified the cause of the memory usage, is it expected for storing the complete Jacobian matrix to require such a significant amount of memory? I noticed the option to set `auto_preconditioning=false` to disable SMP, but according to MOOSE's description in the Steady executioner, this might lead to incorrect computation results, correct? 

> LINEAR will use PETSc control parameter -ksp_only to set the type of SNES for solving the linear system. Note that it only works when you have an exact Jacobian because it is not activating matrix-free calculations.
> 
> NEWTON means PETSc will use the Jacobian provided by kernels (typically not exact) to do the Krylov solve. If the Jacobian is not exact, Newton update in [Eq. (4)](https://mooseframework.inl.gov/source/executioners/Steady.html#moose-equation-f5870d36-ddcc-4f5a-b86d-98bb86ebe1f7) will not reduce the residual effectively and typically results into an unconverged Newton iteration.

I vaguely remember seeing in MOOSE that it's possible to set the off-diagonal level of the Jacobian matrix to determine the range of off-diagonal terms that need to be stored.

**Reply by GiudGiud** [reply]
This may be the problem then; storing the Jacobian.
The next thing you can do to make sure is memory profiling
https://mooseframework.inl.gov/application_development/profiling.html

**Reply by GiudGiud** [reply]
If the Jacobian is incomplete, how are you going to use the LINEAR solve? Or are you planning to go back to PJFNK or a variant?

I think the state of the art for DFEM-SN is sweeping methods btw. These form local systems that are much smaller

**Reply by hityyds** [reply]
Actually, this is a point that has puzzled me. I presumed it directly solves Ax=b using the Krylov method, rather than Jx=R. Consequently, it shouldn't necessitate a Jacobian matrix. I attempted to search for information on the `-ksp_only` parameter of Petsc via Google but couldn't find any pertinent documentation.

**Reply by hityyds** [reply]
I'm also interested in the sweeping method, but that's part of my future plans. Currently, I don't have the capacity to implement a sweeping solver.

**Reply by GiudGiud** [reply]
> LINEAR will use PETSc control parameter -ksp_only to set the type of SNES for solving the linear system. Note that it only works when you have an exact Jacobian because it is not activating matrix-free calculations.

this is the only documentation on our side. This is basically pointing you to PETSC ksp_only indeed. 
Let me know what you find on the petsc side and I ll add a little more on our side on what this actually does

**Reply by hityyds** [reply]
Additionally, may I ask you another question? I've seen that you have responded to similar error messages before. In the last two days, I consistently encountered an error in a postprocessor I wrote myself: "Coupled variable 'psi_0_0' was not found." However, I'm quite certain that I have the variable 'psi_0_0'. I'm calling this postprocessor in an action, and I've double-checked that I initialize it only after declaring all the variables.

**Reply by hityyds** [reply]
> Let me know what you find on the petsc side and I ll add a little more on our side on what this actually does

Currently, I haven't found any helpful information. If I come across other resources later, I'll update you here.

**Reply by GiudGiud** [reply]
If it s a postprocessor then it is a not a (field) variable.
This error message is asking for a variable with that name. If you want to be able to specify either a variable or a postprcoessor, you need to use the functor interfaces. (functors are a base class for both variables and postprocessors)

**Reply by hityyds** [reply]
> If it s a postprocessor then it is a not a (field) variable.

I might not fully understand your response.
'psi_0_0' is a nonlinear variable that I declared in the action. When I disable this postprocessor, I can find it in the output information:
```
Nonlinear System:
  Num DOFs:                23125248
  Num Local DOFs:          937728
  Variables:               { "psi_0_0" "psi_0_1"
```

**Reply by GiudGiud** [reply]
oh I misread, disregard the previous message.

How did you add the parameter where you pass the variable? If you dont use `params.addCoupledVar` then the variable would not be in the list of coupled variables for the postprocessor

**Reply by hityyds** [reply]
I utilize  `params.addRequiredCoupledVar("coupled_variables", "The coupled variables")` to add this parameter and use `coupledComponents("coupled_variables")` and `&coupledValue("coupled_variables", i)` to access these coupled variables.

**Reply by GiudGiud** [reply]
is psi_0 an array variable?

**Reply by hityyds** [reply]
No, it is not. It's just a basic FEM variable.

**Reply by hityyds** [reply]
I declare variables one by one using the action, without employing any techniques like arrays.

**Reply by GiudGiud** [reply]
can you paste the entire postprocessor code? I dont see what could be wrong 
and also what you have in the input for that postprocessor

**Reply by hityyds** [reply]
the relative code are as fellows:
```
//* This file is part of the MOOSE framework
//* https://www.mooseframework.org
//*
//* All rights reserved, see COPYRIGHT for full restrictions
//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
//*
//* Licensed under LGPL 2.1, please see LICENSE for details
//* https://www.gnu.org/licenses/lgpl-2.1.html

#pragma once

#include "ElementPostprocessor.h"

/**
 * This postprocessor computes a volume integral of the specified source.
 *
 * Note that specializations of this integral are possible by deriving from this
 * class and overriding computeQpIntegral().
 */
class ElementIntegralSource : public ElementPostprocessor
{
public:
  static InputParameters validParams();

  ElementIntegralSource(const InputParameters & parameters);

  virtual void initialize() override;
  virtual void execute() override;
  virtual void threadJoin(const UserObject & y) override;
  using Postprocessor::getValue;
  virtual Real getValue() const override;
  virtual void finalize() override;

protected:
  virtual Real computeQpIntegral();
  virtual Real computeIntegral();

  unsigned int _qp;

  Real _integral_value;

  const unsigned int _n_values;  // the number of coupled psis

  std::vector<unsigned int> _var_ids; // it's one unkowns's unique id in all unkowns, and
                                            // each id represent a pointed psi
  std::vector<const VariableValue *> _values; // psi's values
};

InputParameters
ElementIntegralSource::validParams()
{
  InputParameters params = ElementPostprocessor::validParams();
  params.addRequiredCoupledVar("coupled_variables", "The coupled variables");
  return params;
}

ElementIntegralSource::ElementIntegralSource(const InputParameters & parameters)
  : ElementPostprocessor(parameters), _qp(0), _integral_value(0), 
    _n_values(coupledComponents("coupled_variables"))
{
  for (unsigned int i = 0; i < _n_values; ++i)
  {
    _var_ids.push_back(getVar("coupled_variables", i)->number());
    _values.push_back(&coupledValue("coupled_variables", i));
  }
}

void SphinxAction::addPostprocessors()
{
  if (_calc_source_on)
  {
    InputParameters params = _factory.getValidParams("ElementIntegralSource");
    std::vector<VariableName> Coupled_Var_name(_num_Variables);
    for (unsigned int i = 0; i < _num_Variables; i++)
    {
      Coupled_Var_name[i] = _variable_name[i];
    }
    params.set<std::vector<VariableName>>("coupled_variables") = Coupled_Var_name;
    _problem->addPostprocessor("ElementIntegralSource", "total_source", params);
  }
}
```

**Reply by GiudGiud** [reply]
I dont see anything obvious. You did check that addPP is called after the variables were added?
Maybe the `Debug/show_actions` input file option will help check

**Reply by hityyds** [reply]
addPP is called at the end of the action. I will try `Debug/show_actions`. Thanks for your attention.

**Reply by lindsayad** [reply]
> Actually, this is a point that has puzzled me. I presumed it directly solves Ax=b using the Krylov method, rather than Jx=R. 

What is the difference, other than different letters? Probably more to the essence of your question: MOOSE always assumes a nonlinear problem (although this will be changed in the future, see https://github.com/idaholab/moose/issues/25722). If you know your problem is linear, then you can can use the `LINEAR` solve option to avoid some unnecessary computations like a residual evaluation after the linear solve.

> Consequently, it shouldn't necessitate a Jacobian matrix. 

It requires an operator. You can always formulate your linear problem as a nonlinear one and then you can call A J. Or thinking in the reverse, if you know your nonlinear problem happens to be linear, then J delta u = - R ==> A x = b, where J = A, delta u = x, b = -R. And then if your initial guess for u was the zero vector, then it's even simpler with u = x.

> I attempted to search for information on the -ksp_only parameter of Petsc via Google but couldn't find any pertinent documentation.

That documentation should be fixed. The option is rather `-snes_type ksponly`. Here is a PETSc manual page about it https://petsc.org/release/manualpages/SNES/SNESKSPONLY/

**Reply by lindsayad** [reply]
Maybe you're asking if MOOSE can solve linear problems Ax = b, where b does not correspond to a -R coming from a Newton's method, e.g. where b truly only corresponds to terms that do not involve the degrees of freedom, and x corresponds to u and not delta u. The answer to that is no, although this will also be addressed by #25722

**Reply by lindsayad** [reply]
Regarding the memory usage, I would be interested to see the result of memory profiling to see how much of the memory comes from the matrix storage vs. the setup/application of the preconditioner

**Reply by hityyds** [reply]
@GiudGiud  I have identified the reason for the program error: it was indeed a problem with the declaration order. I forgot that in the action, the program determines the execution order based on '_current_task'. This caused pp to be declared before var. I sincerely apologize for the oversight, and I appreciate your valuable time being spent due to my carelessness.

Furthermore, I attempted to disable SMP when using LINEAR. The memory consumption of MOOSE is significantly lower than using PJFNK, and the results of both methods remain consistent.

**Reply by hityyds** [reply]
@lindsayad I encountered issues while installing gperftools for MOOSE, and I will promptly update here if there are any new developments. Thank you very much for your response.

**Reply by lindsayad** [reply]
> Furthermore, I attempted to disable SMP when using LINEAR. The memory consumption of MOOSE is significantly lower than using PJFNK

How did you do this?

**Reply by hityyds** [reply]
Just by setting `auto_preconditioning=false`.

**Reply by lindsayad** [reply]
This is interesting. Changing that option does not change the PETSc solver settings one bit, but in your case dramatically reduces memory usage. In your application, do you have off-diagonal coupling? If so, I would think that your result with `solve_type = LINEAR` would be incorrect with `auto_preconditioning = false`. The fact that you say

> the results of both methods remain consistent

makes me think that you don't have off-diagonal coupling?

**Reply by hityyds** [reply]
Yes, that's correct. I don't have any off-diagonal elements. However, I am curious about why the preconditioner seems to influence the computed results. Isn't a preconditioner supposed to impact only the convergence speed, rather than the actual results? Taking the example of Ax=b with a right preconditioner: the transformed equation is AM^(−1)Mx=b. Regardless of the preconditioner M, it should be equivalent to the original equation.

**Reply by lindsayad** [reply]
The `Preconditioning` block name is probably a little misleading for any solve type other than `PJFNK`. For `NEWTON` the matrix built through the `computeQpJacobian`, etc. routines is used as the Jacobian matrix and for building the preconditioner. For `LINEAR`, it's the same: the matrix is used both for A and for building the preconditioning action of M

**Reply by lindsayad** [reply]
If A is wrong for `solve_type = LINEAR`, then you will get wrong results (scenario 1). Whereas if A was correct and M was "wrong" you would still get the correct answer but as you say, the solver efficiency would not be as high (scenario 2). Since we build A and M from the same matrix, you would be in scenario 1 instead of 2 if the matrix is wrong

**Reply by hityyds** [reply]
Thank you for your detailed explanation! Your clarification on the impact of matrix correctness for different solve types is very helpful. I appreciate your time and insights. I had always thought that computeQpJacobian was only used for computing the preconditioner, as I recall that the computeQpOffDiagJacobian function is allowed to be left undefined.

**Reply by hityyds** [reply]
I am curious if there is a way to directly invoke KSP (bypassing SNES and its Newton methods) to solve the problem? For example, by modifying a small portion of MOOSE's code?

**Reply by GiudGiud** [reply]
There's this, through SNES.
https://petsc.org/release/manualpages/SNES/SNESKSPONLY/
which is what we use when you pass a solve_type of LINEAR
```
    case Moose::ST_LINEAR:
      setSinglePetscOption("-snes_type", "ksponly");
      setSinglePetscOption("-snes_monitor_cancel");
```

**Reply by hityyds** [reply]
Many thanks for your assistance! Your help means a lot to me.
